R/backport_linter.R

```diff
@@ -0,0 +1,82 @@
+#' @describeIn linters that checks for usage of unavailable functions. Not reliable for testing r-devel dependencies.
+#' @export
+backport_linter <- function(r_version = getRversion()) {
+  function(source_file) {
+    if (inherits(r_version, "numeric_version")) r_version <- format(r_version)
+    if (is.null(source_file$xml_parsed_content) || r_version >= "r") return(list())
+
+    xml <- source_file$xml_parsed_content
+
+    names_xpath <- "//*[self::SYMBOL or self::SYMBOL_FUNCTION_CALL]"
+    all_names_nodes <- xml2::xml_find_all(xml, names_xpath)
+    all_names <- xml2::xml_text(all_names_nodes)
+
+    # guaranteed to include 1 by early return above; which.min fails if all TRUE (handled by nomatch)
+    needs_backport_names <- backports[1:(match(FALSE, r_version < names(backports), nomatch = length(backports)) - 1L)]
+
+    # not sapply/vapply, which may over-simplify to vector -- cbind makes sure we have a matrix so rowSums works
+    needs_backport <- do.call(cbind, lapply(needs_backport_names, function(nm) all_names %in% nm))
+    bad_idx <- rowSums(needs_backport) > 0L
+
+    lapply(which(bad_idx), function(ii) {
+      node <- all_names_nodes[[ii]]
+      line1 <- xml2::xml_attr(node, "line1")
+      col1 <- as.integer(xml2::xml_attr(node, "col1"))
+      if (xml2::xml_attr(node, "line2") == line1) {
+        col2 <- as.integer(xml2::xml_attr(node, "col2"))
+      } else {
+        col2 <- nchar(source_file$lines[line1])
+      }
+      Lint(
+        filename = source_file$filename,
+        line_number = as.integer(line1),
+        column_number = col1,
+        type = "warning",
+        message = sprintf(
+          "%s (R %s) is not available for dependency R >= %s.",
+          all_names[ii], names(needs_backport_names)[which(needs_backport[ii, ])], r_version
+        ),
+        line = source_file$lines[[line1]],
+        ranges = list(c(col1, col2)),
+        linter = "backport_linter"
+      )
+    })
+  }
+}
+
+backports <- list(
+  `r` = c("...names", "checkRdContents", "numToBits", "numToInts", "packBits"),
```

Comment:
I think this needs to be `r-devel`

Reply:
(or maybe just "devel"). The tests depend on the exact name and the lint message looks best with `devel` imo:

```
...names (R devel) is not available for dependency R >= 3.0.0.
```